### ISSUE TITLE
Poller should properly handle NRP API failed responses with HttpStatusCode == 200

### ISSUE THREAD

**Issue by kaiminz-db** [Azure/azure-sdk-for-go#23349]
### Bug Report
_Poller should handle NRP API failed responses with HttpStatusCode == 200 OK and allow retries, instead of caching the response indefinitely._

We've noticed GET NetworkOperation polling API calls can return error responses but with status code == 200.
e.g. 

> 	GET https://management.azure.com/subscriptions/{SUB_ID}/providers/Microsoft.Network/locations/westeurope/operations/{OP_ID}
	--------------------------------------------------------------------------------
	RESPONSE 200: 200 OK
	ERROR CODE: **InternalServerError**

Currently Golang SDK treats this as a terminal state and caches the response, thus prevents all retries, although such errors could transient on NRP.

We also contacted Azure support and here is the response

> if SDK expects for polling call to return HttpStatusCode InternalServerError when operation fails, they should use Location URL (not Azure-AsyncOperation) for polling

**Comment by tadelesh** [comment]
could you provide the go code you used to call the nrp api?

**Comment by kaiminz-db** [comment]
> could you provide the go code you used to call the nrp api?

Yes. We noticed this issue happening most when deleting ASGs. Our pseudocode looks like this

```
import (
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
)

req := runtime.NewRequest(http.MethodDelete, "path-for-asg-to-delete")
resp := armClient.Pipeline().Do(req)
poller := runtime.NewPoller(resp, armClient.Pipeline(). &runtime.NewPollerOptions[ApplicationSecurityGroupsClientDeleteResponse]{
			FinalStateVia: runtime.FinalStateViaLocation,
		})

while !poller.Done() {
  ...
  poller.Poll()
  // wait after X minutes 
  ...
} 

result, err := poller.Result()
... 
```

**Comment by tadelesh** [comment]
@jhendrixMSFT the customer is using the raw request call, could you help to identify if it is a core issue?

**Comment by jhendrixMSFT** [comment]
@kaiminz-db would you be able to provide the complete [logs](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore#hdr-Built_in_Logging) of the HTTP requests/responses when you observe this? It would really help narrow down where the problem is.

**Comment by kaiminz-db** [comment]
> @kaiminz-db would you be able to provide the complete [logs](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore#hdr-Built_in_Logging) of the HTTP requests/responses when you observe this? It would really help narrow down where the problem is.

@jhendrixMSFT : we do not have it enabled right now. Let me try to enable the logging and report back if we capture any detailed logs. Thanks!

**Comment by dipan-db** [comment]
> @kaiminz-db would you be able to provide the complete [logs](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore#hdr-Built_in_Logging) of the HTTP requests/responses when you observe this? It would really help narrow down where the problem is.

Thanks. Here is the captured log of this issue. The error happens on polling LRO operation status of delete ASG operation. Below is the capture request and response. Please let me know if you need any other information. Thanks


```
E0730 04:11:44.555546       1 azure_network_security_group_manager.go:1613] "failed to delete asg, Name= i0630001706-7683ts1g-dbasg." err=<
	GET https://management.azure.com/subscriptions/7bb22f4e-f3c2-4bf6-b50e-0769032e07fe/providers/Microsoft.Network/locations/westus/operations/04abc454-0e69-40de-9e1f-3db8ea3dda92
	--------------------------------------------------------------------------------
	RESPONSE 200: 200 OK
	ERROR CODE: InternalServerError
	--------------------------------------------------------------------------------
	{
	  "status": "Failed",
	  "error": {
	    "code": "InternalServerError",
	    "message": "An error occurred.",
	    "details": []
	  }
	}
	--------------------------------------------------------------------------------
 > Name="i0630001706-7683ts1g-dbasg" retried=16
```


```
The corresponding ASG resource in above error is in "provisioningState": "Failed" state
{
    "name": "i0630001706-7683ts1g-dbasg",
    "id": "/subscriptions/7bb22f4e-f3c2-4bf6-b50e-0769032e07fe/resourceGroups/prod-azure-wu-nephos-0-vpo6h1/providers/Microsoft.Network/applicationSecurityGroups/i0630001706-7683ts1g-dbasg",
    "etag": "W/\"b4dc9551-a1d5-4661-b33d-549130e1bddc\"",
    "type": "Microsoft.Network/applicationSecurityGroups",
    "location": "westus",
    "tags": {
        "dblet-lastModifiedTimestamp": "2024-06-30T00:17:07Z"
    },
    "properties": {
        "provisioningState": "Failed"
    }
}
```

**Comment by jhendrixMSFT** [comment]
Thanks this is helpful. If possible, it would also be helpful to provide the initial response to the `DELETE` request as the returned body/headers are used to determine which kind of polling strategy the SDK should use.

From what I see here, the service has returned the status as `Failed` which is one of the three terminal states (`Succeeded`, `Failed`, `Canceled` are the three terminal states). Once an LRO reaches a terminal state, it's considered to be complete. Service-specified errors codes are _never_ consulted when determining if the LRO has reached a terminal state. There's more info on LROs [here](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md) if you're interested.

If the service team intends this to be a transient error and the operation is still in progress on the service side, then it shouldn't be returning a terminal state.
